pkg/controller/plan/adapter/openstack/builder.go

```diff
@@ -843,325 +853,370 @@ func (r *Builder) PodEnvironment(_ ref.Ref, _ *core.Secret) (env []core.EnvVar,
+			r.Log.Info("the image is not ready yet", "image", image.Name)
+			continue
 		}
 
-		if snapshot.Status != "available" {
-			r.Log.Info("Snapshot not ready yet, recheking...", "snapshot", snap.Name)
-			return false, nil
+		originalVolumeDiskId := image.Name
+		if imageProperty, ok := image.Properties[forkliftPropertyOriginalVolumeID]; ok {
+			originalVolumeDiskId = imageProperty.(string)
 		}
-	}
 
-	var vollist []volumes.Volume
-	for _, snap := range snaplist {
-		imageName := fmt.Sprintf("%s-%s", r.Migration.Name, snap.VolumeID)
-		pager := volumes.List(client.blockStorageService, volumes.ListOpts{
-			Name:  imageName,
-			Limit: 1,
-		})
-		pages, err := pager.AllPages()
-		if err != nil {
-			return true, err
-		}
-		isEmpty, err := pages.IsEmpty()
+		_, err = r.getVolumePopulator(image.Name)
 		if err != nil {
-			return true, err
-		}
-		if !isEmpty {
-			vols, err := volumes.ExtractVolumes(pages)
+			if !k8serr.IsNotFound(err) {
+				err = liberr.Wrap(err)
+				return
+			}
+
+			var populatorName string
+			populatorName, err = r.createVolumePopulator(image, secretName, vmRef.ID)
 			if err != nil {
-				return true, err
+				err = liberr.Wrap(err)
+				return
 			}
-			vollist = append(vollist, vols...)
 
-			continue
-		}
+			storageClassName := r.Context.Map.Storage.Spec.Map[0].Destination.StorageClass
+			volumeType := r.getVolumeType(workload, originalVolumeDiskId)
+			if volumeType != "" {
+				storageClassName, err = r.getStorageClassName(workload, volumeType)
+				if err != nil {
+					err = liberr.Wrap(err)
+					return
+				}
+			}
+			accessModes, volumeMode, err = r.getVolumeAndAccessMode(storageClassName)
+			if err != nil {
+				err = liberr.Wrap(err)
+				return
+			}
 
-		volume, err := volumes.Create(client.blockStorageService, volumes.CreateOpts{
-			Name:        imageName,
-			SnapshotID:  snap.ID,
-			Size:        snap.Size,
-			Description: imageName,
-		}).Extract()
-		if err != nil {
-			err = liberr.Wrap(
-				err,
-				"Failed to create volume from snapshot.",
-				"volumeID",
-				volume.ID)
-			return true, err
+			pvc := r.persistentVolumeClaimWithSourceRef(image, storageClassName, populatorName, accessModes, volumeMode, annotations)
+			if pvc == nil {
+				err = liberr.New("couldn't build the PVC",
+					"image", image.Name, "storageClassName", storageClassName, "populatorName", populatorName)
+				return
+			}
+			err = r.Client.Create(context.TODO(), pvc, &client.CreateOptions{})
+			if err != nil {
+				if !k8serr.IsAlreadyExists(err) {
+					err = liberr.Wrap(err)
+					return
+				}
+				err = nil
+				continue
+			}
+			pvcNames = append(pvcNames, pvc.Name)
 		}
-		vollist = append(vollist, *volume)
 	}
+	return
+}
 
-	for _, vol := range vollist {
-		volume, err := volumes.Get(client.blockStorageService, vol.ID).Extract()
-		if err != nil {
-			return true, err
+func (r *Builder) createVolumePopulator(image *model.Image, secretName, vmId string) (name string, err error) {
```

Comment:
It would be nice to have `CR` in the method name to know we create here CR and not a PVC.